Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertMaterialQuantities.cs

```diff
@@ -11,69 +10,148 @@ namespace Objects.Converter.Revit
+    }
 
-      var speckleMaterial = ConvertAndCacheMaterial(material.Id, material.Document);
-      var materialQuantity = new Objects.Other.MaterialQuantity(speckleMaterial, volume, area, units);
+    private MaterialQuantity Create(
+      Element element,
+      ElementId materialId,
+      double areaRevitInternalUnits,
+      double volumeRevitInternalUnits,
+      string units)
+    {
+      Other.Material speckleMaterial = ConvertAndCacheMaterial(materialId, element.Document);
+      double factor = ScaleToSpeckle(1);
+      double area = factor * factor * areaRevitInternalUnits;
+      double volume = factor * factor * factor * volumeRevitInternalUnits;
+      MaterialQuantity materialQuantity = new(speckleMaterial, volume, area, units);
 
       if (LocationToSpeckle(element) is ICurve curve)
+      {
         materialQuantity["length"] = curve.length;
+      }
+      else if (element is DB.Architecture.Railing)
+      {
+        materialQuantity["length"] = (element as DB.Architecture.Railing).GetPath().Sum(e => e.Length) * factor;
+      }
+      else if (element is DB.Architecture.ContinuousRail)
+      {
+        materialQuantity["length"] = (element as DB.Architecture.ContinuousRail).GetPath().Sum(e => e.Length) * factor;
+      }
+
       return materialQuantity;
     }
 
-    #endregion
-
-    #region MaterialQuantities
-    public IEnumerable<Objects.Other.MaterialQuantity> MaterialQuantitiesToSpeckle(DB.Element element, string units)
+    private (double, double) GetAreaAndVolumeFromSolids(List<Solid> solids, ElementId? materialId = null)
     {
-      
-      var matIDs = element?.GetMaterialIds(false); 
-      // Does not return the correct materials for some categories
-      // Need to take different approach for MEP-Elements
-      if (matIDs == null || !matIDs.Any() &&  element is MEPCurve)
+      if (materialId != null)
       {
-        DB.Material mepMaterial = ConverterRevit.GetMEPSystemRevitMaterial(element);
-        if (mepMaterial != null) matIDs.Add(mepMaterial.Id);
+        solids = solids
+          .Where(
+            solid => solid.Volume > 0
+            && !solid.Faces.IsEmpty
+            && solid.Faces.get_Item(0).MaterialElementId == materialId)
+          .ToList();
       }
 
-      if (matIDs == null || !matIDs.Any())
-        return null;
+      double volume = solids.Sum(solid => solid.Volume);
+      IEnumerable<double> areaOfLargestFaceInEachSolid = solids
+          .Select(solid => solid.Faces.Cast<Face>().Select(face => face.Area)
+          .Max());
+      double area = areaOfLargestFaceInEachSolid.Sum();
+      return (area, volume);
+    }
 
-      var materials = matIDs.Select(material => element.Document.GetElement(material) as DB.Material);
-      return MaterialQuantitiesToSpeckle(element, materials, units);
+    private IEnumerable<DB.ElementId> GetMaterialsFromSolids(List<Solid> solids)
+    {
+      return solids
+        .Where(solid => solid.Volume > 0 && !solid.Faces.IsEmpty)
```

Comment:
I guess if !solid.Faces.IsEmpty is true, the m.faces.get_item(0) should be valid? Can there be other faces that should be considered?

Reply:
This is an assumption that comes from the Gauip dev that a single solid will only have one material. I'm actually not sure if this is true.